include/libvfio-user.h

```diff
@@ -409,23 +409,34 @@ typedef struct vfu_dma_info {
  * Called when a guest registers one of its DMA regions via a VFIO_USER_DMA_MAP
  * message.
  *
+ * If the callback returns -1 and errno is set to EBUSY then
+ * libvfio-user does not respond to the client and does not process new
+ * messages until vfu_async_done is called by the device, effectively making
+ * the DMA map operation asynchronous.
+ *
  * @vfu_ctx: the libvfio-user context
  * @info: the DMA info
+ *
+ * @returns: 0 on success, -1 on failure with errno set.
  */
-typedef void (vfu_dma_register_cb_t)(vfu_ctx_t *vfu_ctx, vfu_dma_info_t *info);
+typedef int (vfu_dma_register_cb_t)(vfu_ctx_t *vfu_ctx, vfu_dma_info_t *info);
 
 /*
- * Function that is called when the guest unregisters a DMA region. The device
- * must release all references to that region before the callback returns.
- * This is required if you want to be able to access guest memory directly via
- * a mapping.
- *
- * The callback should return 0 on success, -1 with errno set on failure
+ * Function that is called when the guest unregisters a DMA region.  The
+ * callback should return 0 on success and -1 on failure setting errno.
  * (although unregister should not fail: this will not stop a guest from
- * unregistering the region).
+ * unregistering the region).  This callback is required if you want to be able
+ * to access guest memory directly via a mapping.
+ *
+ * The callback can be asynchronous, just like vfu_dma_register_cb_t.
+ *
+ * The device must release all references to that region before the callback
```

Comment:
Should specify this is the case only if sync, and if async, then it must be true when calling vfu_async_done() ?